Standardize and add SSL settings

edmocosta opened the pull request:
### What this PR does?
Added the following SSL settings:
`ssl_enabled`: Enable/disable the SSL settings. Infer the value from the hosts' scheme if neither the deprecate :ssl nor the new :ssl_enabled configs were set
`ssl_certificate`: OpenSSL-style X.509 certificate file to authenticate the client
`ssl_key`: OpenSSL-style RSA private key that corresponds to the `ssl_certificate`
`ssl_truststore_path`: he JKS truststore to validate the server's certificate
`ssl_truststore_type`: The format of the truststore file
`ssl_truststore_password`: The truststore password
`ssl_keystore_path`: The keystore used to present a certificate to the server
`ssl_keystore_type`: The format of the keystore file
`ssl_keystore_password`: The keystore password
`ssl_cipher_suites`: The list of cipher suites to use
`ssl_supported_protocols`: Supported protocols with versions

Reviewed and deprecated SSL settings to comply with Logstash's naming convention
`ssl` in favor of `ssl_enabled`: 
`ca_file` in favor of `ssl_certificate_authorities`
`ssl_certificate_verification` in favor of `ssl_verification_mode`


The behavior standardization across plugins, such as the accepted certificate formats, default values, etc will be tackled in future PRs.

---
Closes https://github.com/elastic/logstash/issues/14922
Closes https://github.com/logstash-plugins/logstash-input-elasticsearch/issues/96
Closes https://github.com/logstash-plugins/logstash-input-elasticsearch/issues/71
Closes https://github.com/logstash-plugins/logstash-input-elasticsearch/issues/42
Closes https://github.com/logstash-plugins/logstash-input-elasticsearch/issues/115

andsel commented on docs/index.asciidoc:
```diff
@@ -140,7 +153,7 @@ input plugins.
   * There is no default value for this setting.
 
-Authenticate using Elasticsearch API key. Note that this option also requires enabling the `ssl` option.
+Authenticate using Elasticsearch API key. Note that this option also requires enabling the `ssl_enabled` option.
```
```suggestion
Authenticate using Elasticsearch API key. Note that this option also requires enabling the <<plugins-{type}s-{plugin}-ssl_enabled>> option.
```

andsel commented on docs/index.asciidoc:
```diff
@@ -149,6 +162,7 @@ API key API].
 
 [id="plugins-{type}s-{plugin}-ca_file"]
 ===== `ca_file` 
+deprecated[8.8.0, Replaced by <<plugins-{type}s-{plugin}-ssl_certificate_authorities>>]
```
This deprecation version should related the plugin's version and not the Logstash version.
Same for following changes about the `deprecated`

andsel commented on docs/index.asciidoc:
```diff
@@ -408,18 +422,39 @@ instructions into the query.
+
+SSL certificate to use to authenticate the client. This certificate should be an OpenSSL-style X.509 certificate file.
+
+NOTE: This setting can be used only if `ssl_key` is set.
```
```suggestion
NOTE: This setting can be used only if <<plugins-{type}s-{plugin}-ssl_key>> is set.
```

andsel commented on docs/index.asciidoc:
```diff
@@ -408,18 +422,39 @@ instructions into the query.
+  * Value type is a list of <<path,path>>
+  * There is no default value for this setting
+
+The .cer or .pem files to validate the server's certificate.
```
```suggestion
The `.cer` or `.pem` files to validate the server's certificate.
```

andsel commented on docs/index.asciidoc:
```diff
@@ -408,18 +422,39 @@ instructions into the query.
+
+The .cer or .pem files to validate the server's certificate.
+
+NOTE: You cannot use this setting and `ssl_truststore_path` at the same time.
```
```suggestion
NOTE: You cannot use this setting and <<plugins-{type}s-{plugin}-ssl_truststore_path>> at the same time.
```

andsel commented on docs/index.asciidoc:
```diff
@@ -429,6 +464,120 @@ man-in-the-middle or other compromised infrastructure.
+  * Value type is <<path,path>>
+  * There is no default value for this setting.
+
+OpenSSL-style RSA private key that corresponds to the `ssl_certificate`.
```
```suggestion
OpenSSL-style RSA private key that corresponds to the <<plugins-{type}s-{plugin}-ssl_certificate>>.
```

andsel commented on docs/index.asciidoc:
```diff
@@ -429,6 +464,120 @@ man-in-the-middle or other compromised infrastructure.
+  * There is no default value for this setting.
+
+The keystore used to present a certificate to the server.
+It can be either .jks or .p12
```
```suggestion
It can be either `.jks` or `.p12`
```

andsel commented on docs/index.asciidoc:
```diff
@@ -429,6 +464,120 @@ man-in-the-middle or other compromised infrastructure.
+
+List of allowed SSL/TLS versions to use when establishing a connection to the Elasticsearch cluster.
+
+For Java 8 `'TLSv1.3'` is supported  only since **8u262** (AdoptOpenJDK), but requires that you set the
```
```suggestion
For Java 8 `'TLSv1.3'` is supported only since **8u262** (AdoptOpenJDK), but requires that you set the
```

andsel commented on docs/index.asciidoc:
```diff
@@ -429,6 +464,120 @@ man-in-the-middle or other compromised infrastructure.
+  * Value type is <<password,password>>
+  * There is no default value for this setting.
+
+Set the truststore password
```
```suggestion
Set the truststore password.
```

andsel commented on lib/logstash/inputs/elasticsearch.rb:
```diff
@@ -185,15 +188,60 @@ class LogStash::Inputs::Elasticsearch < LogStash::Inputs::Base
+  config :ssl_certificate_verification, :validate => :boolean, :default => true, :deprecated => "Set 'ssl_verification_mode' instead."
+
+  # Options to verify the server's certificate.
+  # "full": validates that the provided certificate has an issue date that’s within the not_before and not_after dates;
+  # chains to a trusted Certificate Authority (CA); has a hostname or IP address that matches the names within the certificate.
+  # "none": performs no certificate validation. Disabling this severely compromises security (https://www.cs.utexas.edu/~shmat/shmat_ccs12.pdf)
+  config :ssl_verification_mode, :validate => %w[full none], :default => 'full'
+
+  # OpenSSL-style X.509 certificate certificate to authenticate the client
+  config :ssl_certificate, :validate => :path
+
+  # OpenSSL-style RSA private key to authenticate the client
+  config :ssl_key, :validate => :path
+
+  # The JKS truststore to validate the server's certificate.
+  # Use either `:ssl_truststore_path` or `:ssl_certificate_authorities`
+  config :ssl_truststore_path, :validate => :path
+
+  # The format of the truststore file. It must be either jks or pkcs12
+  config :ssl_truststore_type, :validate => %w[pkcs12 jks]
+
+  # Set the truststore password
+  config :ssl_truststore_password, :validate => :password
+
+  # The keystore used to present a certificate to the server.
+  # It can be either .jks or .p12
+  config :ssl_keystore_path, :validate => :path
+
+  # The format of the keystore file. It must be either jks or pkcs12
+  config :ssl_keystore_type, :validate => %w[pkcs12 jks]
+
+  # Set the keystore password
+  config :ssl_keystore_password, :validate => :password
+
+  # The list of cipher suites to use, listed by priorities.
+  # Supported cipher suites vary depending on which version of Java is used.
+  config :ssl_cipher_suites, :validate => :string, :list => true
+
+  # Supported protocols with versions.
+  config :ssl_supported_protocols, :validate => %w[TLSv1.1 TLSv1.2 TLSv1.3], :default => [], :list => true
```
For new config settings we should respect lexicographic order, I think.

edmocosta replied:
Is it also valid for the code? I thought it was only for the docs. IMO, keeping related configs together is easier to read & understand than respecting a lexicographic order. I'm fine with rearranging them if that's the style guide.

andsel replied:
In general you are right, but given that all these `ssl_*` stuff is self aggregating, in terms of lexicographical order, for new chucks of settings like this, I think that would be helpfull such order, because permit to locate the settings immediately also in the code base

andsel commented on lib/logstash/inputs/elasticsearch.rb:
```diff
@@ -426,34 +484,116 @@ def validate_authentication
+    if ssl_certificate && ssl_key
+      ssl_options[:client_cert] = ssl_certificate
+      ssl_options[:client_key] = ssl_key
+    elsif !!ssl_certificate ^ !!ssl_key
```
Could use the logical `||` because the `true` && `true` is captured by the other branch.

andsel reviewed (changes requested):
Left some comments, for your evaluation

andsel commented on lib/logstash/inputs/elasticsearch.rb:
```diff
@@ -426,34 +484,120 @@ def validate_authentication
+    ssl_certificate_authorities, ssl_truststore_path, ssl_certificate, ssl_keystore_path = params.values_at('ssl_certificate_authorities', 'ssl_truststore_path', 'ssl_certificate', 'ssl_keystore_path')
+
+    if ssl_certificate_authorities && ssl_truststore_path
+      raise(LogStash::ConfigurationError, 'Use either "ssl_certificate_authorities/ca_file" or "ssl_truststore_path" when configuring the CA certificate')
```
The preferred style to raise an exception es per this [style guide](https://github.com/rubocop/ruby-style-guide#exceptions)
Is to use
```ruby
raise LogStash::ConfigurationError, 'Use either "ssl_certificate_authorities/ca_file" or "ssl_truststore_path" when configuring the CA certificate'
```
or 
```ruby
raise LogStash::ConfigurationError.new 'Use either "ssl_certificate_authorities/ca_file" or "ssl_truststore_path" when configuring the CA certificate'
```

I wouldn't update the existing invocations in PR, just the new insertions, also if it generate style inconsistency.

andsel reviewed (changes requested):
We are almost at the finish just a style nitpick

edmocosta commented:
> Left some comments, for your evaluation

Thank you again for reviewing it, @andsel. I've addressed your suggestions plus the [suggestion #1](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1118#discussion_r1131890635 ) and [suggestion #2](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1118#discussion_r1131892391) made on the output plugin.

andsel reviewed (approved):
LGTM

karenzone commented on docs/index.asciidoc:
```diff
@@ -463,6 +575,55 @@ option when authenticating to the Elasticsearch server. If set to an
 
+[id="plugins-{type}s-{plugin}-deprecated-options"]
+==== Elasticsearch Input deprecated configuration options
+
```
This is a departure from the standard approach we've used for deprecated options in other plugins. For example, https://www.elastic.co/guide/en/logstash/current/plugins-inputs-beats.html#plugins-inputs-beats-options.  

This approach is cleaner, but not as discoverable for users with older configs.
